### PR TITLE
rbac: relax RBAC requirements for `EXPLAIN <query>`

### LINKED ISSUE #20478: Enable `EXPLAIN` for instrospection user
### Feature request

The field engineering team is using teleport for troubleshooting customer environments without access to customer data. It's a valuable tool that avoids sometimes tedious back and forth with customers. However, it's currently not possible to retrieve the explain plan of a query through that mechanism.

```
localhost > explain SELECT * FROM auctions LIMIT 1;
ERROR:  permission denied for SOURCE "materialize.qck.auctions"
```
It would be great to extend the permissions of the introspection users to enable explain statements. Nikhil does [not have any security related concerns](https://materializeinc.slack.com/archives/C04R299RJ7R/p1689087263002679).

### PR DESCRIPTION
**Pull request by aalexandrov**
At the moment, executing `EXPLAIN <query>` statements requires read privileges to all items referenced in the query.

Since `EXPLAIN` is not evaluating a query (except for partially evaluating constant paths), we can relax the RBAC requirements to only require `USAGE` privilege on the schema of items contained in the query.

### Motivation

* This PR adds a known-desirable feature.

Closes #20478.

### Tips for reviewer

1. I did `--rewrite-tests` on tests containing RBAC expressions and got these changes. Looking at the other tests in `privilege_checks.slt`, I don't see any other passing tests, so I'm not sure if I should keep the rewritten items or delete them.
2. I think that this fixes the issue with `EXPLAIN <query>`, but not with `EXPLAIN <timestamp>`. I don't think that the latter will reveal more customer information (I would consider the progress timestamps metadata), so I would advocate for fixing this in the same PR as well.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: test/sqllogictest/privilege_checks.slt [hunk 1]
```diff
@@ -1812,12 +1830,18 @@ COMPLETE 0
...
 simple conn=child,user=child
 EXPLAIN SELECT * FROM v;
 ----
-db error: ERROR: permission denied for TABLE "materialize.public.t"
+Explained Query:
+  ReadStorage materialize.public.t
+
+COMPLETE 1
 
 simple conn=mz_system,user=mz_system
```
Comment (jkosh44) [inline test/sqllogictest/privilege_checks.slt:1846]: You can remove this grant and all revokes, except `REVOKE CREATE ON SCHEMA materialize.public FROM view_owner;`

**Review by jkosh44 — APPROVED** [review]
LGTM, but I think the test needs a bit of a cleanup.

**Comment by jkosh44** [comment]
> so I would advocate for fixing this in the same PR as well.

I would be OK with that.

**Comment by aalexandrov** [comment]
@jkosh44 I did all suggestions and changed `EXPLAIN TIMESTAMP` accordingly. Now waiting for CI to tell me if there are any tests that I should fix regarding that.

File: test/sqllogictest/privilege_checks.slt [hunk 2]
```diff
@@ -1779,68 +1777,35 @@ GRANT USAGE ON SCHEMA materialize.public TO joe;
...
-COMPLETE 0
+Explained Query:
+  ReadStorage materialize.public.t
 
-simple conn=mz_system,user=mz_system
-REVOKE USAGE ON SCHEMA materialize.public FROM joe;
-----
-COMPLETE 0
+EOF
+COMPLETE 1
```
Comment (jkosh44) [inline test/sqllogictest/privilege_checks.slt:1805]: It looks like these two test cases are identical? I think we can remove one.

**Review by jkosh44 — APPROVED** [review]

File: test/sqllogictest/privilege_checks.slt (hunk 2 above)
Comment (aalexandrov) [reply test/sqllogictest/privilege_checks.slt:1805]: They target different connection and user parameters:

```
simple conn=joe,user=joe
EXPLAIN SELECT * FROM v;
----
db error: ERROR: permission denied for SCHEMA "materialize.public"

simple conn=child,user=child
EXPLAIN SELECT * FROM v;
----
db error: ERROR: permission denied for SCHEMA "materialize.public"
```

**Pull request merged**